File: 6. N 字形变换.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    const q = new Array(numRows).fill(0).map(x => new Array());

    let index = 0;
    while (index < s.length) {

        let qPtr = 0;
        while (index < s.length && qPtr < q.length) {
            q[qPtr].push(s[index]);
            index++;
            qPtr++;
        }
        qPtr = q.length - 2;
        while (index < s.length && qPtr > 0) {
            q[qPtr].push(s[index]);
            index++;
            qPtr--;
        }
    }

    let out = '';
    for (let i = 0; i < q.length; i++) {
        out += q[i].join('');
    }
    return out;
};

File: 6. N 字形变换.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) return s; // 注意这里需要判断行数为1的情况

        vector<string> v(numRows);

        int increment = -1;
        int vPtr = 0;
        for (int i = 0; i < s.length(); i++) {
            v[vPtr] += s[i];
            if (vPtr == 0 || vPtr == numRows - 1) {
                increment = -increment;
            }
            vPtr += increment;
        }

        string out = "";
        for (int i = 0; i < v.size(); i++) {
            out += v[i];
        }
        
        return out;
    }
};